sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -137,8 +138,16 @@ case class EnsureRequirements(
         Some(finalCandidateSpecs.values.maxBy(_.numPartitions))
       }
 
+      // Check if 1) all children are of `DataSourcePartitioning` and 2) they are all compatible
+      // with each other. If both are true, skip shuffle.
+      val allCompatible = childrenIndexes.sliding(2).map {
+            case Seq(a, b) =>
+              checkDataSourceSpec(specs(a)) && checkDataSourceSpec(specs(b)) &&
+                  specs(a).isCompatibleWith(specs(b))
+          }.forall(_ == true)
+
       children = children.zip(requiredChildDistributions).zipWithIndex.map {
-        case ((child, _), idx) if !childrenIndexes.contains(idx) =>
+        case ((child, _), idx) if allCompatible || !childrenIndexes.contains(idx) =>
```

Comment:
This is because we don't allow Spark to re-shuffle the other side of the join using `DataSourceShuffleSpec` yet, so its `canCreatePartitioning` is false. However, we do want to skip shuffle the if all sides of a join all have compatible `DataSourceShuffleSpec`, and so there is a special check here.

Reply:
`DataSourceShuffleSpec` (without enabling it to re-shuffle other children) can't fit nicely into this framework at the moment, I think.

 Since its `canCreatePartitioning` is false, we'll treat it the same as `RangeShuffleSpec`. In case both sides are of `DataSourceShuffleSpec`, the `bestSpecOpt` will be `None` in `EnsureRequirements` and Spark will create default partitioning from the required distribution instead.

Reply:
you are right